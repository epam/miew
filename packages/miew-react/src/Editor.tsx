import useResizeObserver from 'use-resize-observer'
import { AppContainer } from './components/App'

const MEDIA_SIZES = {
  smallWidth: 800,
  smallHeight: 400
}

const Editor = () => {
  const { ref, width, height } = useResizeObserver<HTMLDivElement>()

  const isSizeSmall =
    (height && height <= MEDIA_SIZES.smallHeight) ||
    (width && width <= MEDIA_SIZES.smallWidth)

  const containerStyle = {
    height: '100%',
    width: '100%',
    backgroundColor: isSizeSmall ? '#eee' : '#fff'
  }

  return (
    <div css={containerStyle} ref={ref}>
      <AppContainer />
    </div>
  )
}

export { Editor }
