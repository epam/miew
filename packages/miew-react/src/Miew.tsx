import useResizeObserver from 'use-resize-observer'
import clsx from 'clsx'
import classes from './Miew.module.scss'

const MEDIA_SIZES = {
  smallWidth: 800,
  smallHeight: 400
}

type Props = {
  text: string
}

function Miew({ text }: Props) {
  const { ref, width, height } = useResizeObserver<HTMLDivElement>()

  return (
    <div
      className={clsx(classes.miew, {
        [classes.small]:
          (height && height <= MEDIA_SIZES.smallHeight) ||
          (width && width <= MEDIA_SIZES.smallWidth)
      })}
      ref={ref}
    >
      <p>HELLO, I am Miew!</p>
      <p>{text}</p>
      <p>
        Width: {width} Height: {height}
      </p>
    </div>
  )
}

export { Miew }
