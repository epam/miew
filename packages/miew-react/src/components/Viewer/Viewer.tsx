import { useEffect, useRef } from 'react'
// @ts-ignore
// import Miew from 'miew'

export const Viewer = ({ Miew }) => {
  const miewContainer = useRef<HTMLDivElement>(null)

  useEffect(() => {
    if (!miewContainer.current) return
    const Viewer = new Miew({
      container: miewContainer.current,
      load: '2FD7',
      reps: [
        {
          mode: 'BS'
        }
      ],
      settings: {
        autoPreset: false,
        bg: { color: 0x202020 },
        editing: true,
        inversePanning: true
      }
    })

    if (Viewer.init()) {
      Viewer.run()
    }
  }, [])

  return <div ref={miewContainer}></div>
}
