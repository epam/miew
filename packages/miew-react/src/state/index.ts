export * from './store'
export * from './rootSaga'
export * from './hooks'
