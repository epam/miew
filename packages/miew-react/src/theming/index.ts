export * from './defaultTheme'
export * from './theme.types'
