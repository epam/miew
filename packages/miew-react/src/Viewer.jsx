import Miew from 'miew';
import PropTypes from 'prop-types';
import React, { useEffect, useLayoutEffect, useRef } from 'react';
import * as styles from './Viewer.module.scss';
import 'miew/dist/Miew.css';

/**
 * Creates and initializes a Miew instance
 * @param {Object} miewRef - Ref to store the Miew instance
 * @param {Object} options - Miew initialization options
 * @param {Function} onInit - Callback function called after successful initialization
 * @param {Function} onError - Callback function called when an error occurs
 */
function createMiewRef(miewRef, options, onInit, onError) {
  if (!miewRef.current) {
    try {
      const miew = new Miew(options);
      if (miew.init()) {
        miewRef.current = miew;
        miew.run();

        if (typeof onInit === 'function') {
          onInit(miew);
        }
      } else {
        const error = new Error('Failed to initialize Miew');
        if (typeof onError === 'function') {
          onError(error);
        }
      }
    } catch (error) {
      if (typeof onError === 'function') {
        onError(error);
      }
    }
  }
}

/**
 * Properly destroys a Miew instance and cleans up resources
 * @param {Object} miewRef - Ref containing the Miew instance
 */
function destroyMiewRef(miewRef) {
  if (miewRef.current) {
    miewRef.current.term();
  }
  miewRef.current = null;
}

/**
 * React component wrapper for Miew 3D molecular viewer
 * @param {Object} props - Component props
 * @param {Function} props.onInit - Callback function called after successful Miew initialization
 * @param {Function} props.onError - Callback function called when an error occurs during initialization
 * @param {Object} props.options - Miew configuration options
 * @param {*} props.theme - Legacy prop kept for backwards compatibility, will be removed in future versions
 * @returns {JSX.Element} The Viewer component
 */
export default function Viewer({ onInit, onError, options, theme }) {
  const miewRef = useRef();
  const rootRef = useRef();

  // Deprecation warning for theme prop
  useEffect(() => {
    if (theme !== undefined) {
      console.warn(
        'Warning: The "theme" prop is deprecated and will be removed in future versions of miew-react. It is kept for backwards compatibility only.',
      );
    }
  }, [theme]);

  useLayoutEffect(() => {
    const settings = { axes: false, fps: false, ...options?.settings };
    createMiewRef(miewRef, { ...options, container: rootRef.current, settings }, onInit, onError);
    return () => destroyMiewRef(miewRef);
  }, [onInit, onError, options]);

  return (
    <div className={styles.root} ref={rootRef}>
      Viewer
    </div>
  );
}

Viewer.propTypes = {
  onInit: PropTypes.func,
  onError: PropTypes.func,
  options: PropTypes.shape({
    load: PropTypes.string, // Molecule identifier or file path
    type: PropTypes.string, // File type (pdb, cif, etc.)
    settings: PropTypes.object, // Viewer settings (axes, fps, autoRotation, bg, etc.)
    settingsCookie: PropTypes.string, // Cookie name for storing settings
    cookiePath: PropTypes.string, // Cookie path
    reps: PropTypes.arrayOf(PropTypes.object), // Representations array
    container: PropTypes.instanceOf(typeof HTMLElement !== 'undefined' ? HTMLElement : Object), // DOM container element
  }),
  /** @deprecated This prop is deprecated and will be removed in future versions */
  theme: PropTypes.any,
};

Viewer.defaultProps = {
  onInit: undefined,
  onError: undefined,
  options: undefined,
  theme: undefined,
};
