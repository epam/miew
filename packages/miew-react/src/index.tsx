import { Miew } from './Miew'

export default Miew
