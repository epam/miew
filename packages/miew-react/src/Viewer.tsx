import { useLayoutEffect, useRef } from 'react'
import { Provider } from 'react-redux'
import { Miew, MiewOptions } from 'miew'
import useResizeObserver from 'use-resize-observer'
import { css, Theme, ThemeProvider } from '@emotion/react'
import { createTheme } from '@mui/material/styles'
import { CssBaseline } from '@mui/material'
import { merge } from 'lodash'
import { store } from 'state'
import { defaultTheme, MiewTheme } from './theming'
import hexToRgba from 'hex-to-rgba'

const MEDIA_SIZES = {
  smallWidth: 800,
  smallHeight: 400
}

type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>
}

type ViewerProps = {
  onInit?: (miew: Miew) => void
  options?: MiewOptions
  theme?: DeepPartial<MiewTheme>
}

const muiTheme = createTheme()

const getViewerStyle = (isSizeSmall) => {
  return (theme: Theme) => {
    const palette = theme.miew.palette
    return css({
      backgroundColor: isSizeSmall ? palette.accent.main : palette.primary.main,
      height: '100%',
      width: '100%',
      position: 'relative',
      '& > .miew-canvas': {
        height: '100%',
        width: '100%'
      },
      '& > .overlay': {
        position: 'absolute',
        top: '10px', // TODO: Should be dynamic, depending on mode; refactor when controlPanel is merged
        right: '10px',
        borderRadius: '4px',
        color: palette.secondary.light,
        backgroundColor: hexToRgba(palette.primary.dark, 0.75),
        p: {
          margin: '10px',
          textAlign: 'left'
        }
      }
    })
  }
}

const Viewer = ({ onInit, options, theme }: ViewerProps) => {
  const viewerTheme = theme ? merge(defaultTheme, theme) : defaultTheme

  const ref = useRef<HTMLDivElement>(null)
  const { width, height } = useResizeObserver<HTMLDivElement>({ ref })

  const isSizeSmall =
    (height && height <= MEDIA_SIZES.smallHeight) ||
    (width && width <= MEDIA_SIZES.smallWidth)

  const viewerStyle = getViewerStyle(isSizeSmall)

  useLayoutEffect(() => {
    const miew = new Miew({
      container: ref?.current,
      ...options
    })
    if (miew.init()) miew.run()
    if (typeof onInit === 'function') onInit(miew)
  }, [options, onInit])

  return (
    <Provider store={store}>
      <ThemeProvider theme={merge(muiTheme, { miew: viewerTheme })}>
        <CssBaseline />
        <div ref={ref} css={viewerStyle} />
      </ThemeProvider>
    </Provider>
  )
}

export { Viewer }
