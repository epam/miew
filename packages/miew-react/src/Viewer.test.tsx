import { render } from '@testing-library/react'
import 'jest-canvas-mock'
import { Viewer } from 'src/Viewer'

// TODO create mock folder, and move all mocks there
jest.mock('use-resize-observer', () =>
  jest.fn().mockReturnValue({ ref: null, width: 100, height: 100 })
)

jest.mock('miew', () =>
  jest.fn().mockReturnValue({ init: jest.fn(() => true), run: jest.fn() })
)

describe('Miew component', () => {
  it('should be rendered', () => {
    expect(render(<Viewer />)).toMatchSnapshot()
  })
})
