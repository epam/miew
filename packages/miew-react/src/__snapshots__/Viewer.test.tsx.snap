// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewer component should apply custom theme when theme prop is provided 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  background-color: #f00;
  height: 100%;
  width: 100%;
  position: relative;
}

.emotion-0>.miew-canvas {
  height: 100%;
  width: 100%;
}

.emotion-0>.overlay {
  position: absolute;
  top: 10px;
  right: 10px;
  border-radius: 4px;
  color: #C0C0C0;
  background-color: rgba(142, 67, 67, 0.75);
}

.emotion-0>.overlay p {
  margin: 10px;
  text-align: left;
}

<body>
    <div>
      <div
        class="emotion-0"
      />
    </div>
  </body>,
  "container": .emotion-0 {
  background-color: #f00;
  height: 100%;
  width: 100%;
  position: relative;
}

.emotion-0>.miew-canvas {
  height: 100%;
  width: 100%;
}

.emotion-0>.overlay {
  position: absolute;
  top: 10px;
  right: 10px;
  border-radius: 4px;
  color: #C0C0C0;
  background-color: rgba(142, 67, 67, 0.75);
}

.emotion-0>.overlay p {
  margin: 10px;
  text-align: left;
}

<div>
    <div
      class="emotion-0"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Viewer component should be rendered 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  background-color: #2E6DA4;
  height: 100%;
  width: 100%;
  position: relative;
}

.emotion-0>.miew-canvas {
  height: 100%;
  width: 100%;
}

.emotion-0>.overlay {
  position: absolute;
  top: 10px;
  right: 10px;
  border-radius: 4px;
  color: #C0C0C0;
  background-color: rgba(64, 64, 64, 0.75);
}

.emotion-0>.overlay p {
  margin: 10px;
  text-align: left;
}

<body>
    <div>
      <div
        class="emotion-0"
      />
    </div>
  </body>,
  "container": .emotion-0 {
  background-color: #2E6DA4;
  height: 100%;
  width: 100%;
  position: relative;
}

.emotion-0>.miew-canvas {
  height: 100%;
  width: 100%;
}

.emotion-0>.overlay {
  position: absolute;
  top: 10px;
  right: 10px;
  border-radius: 4px;
  color: #C0C0C0;
  background-color: rgba(64, 64, 64, 0.75);
}

.emotion-0>.overlay p {
  margin: 10px;
  text-align: left;
}

<div>
    <div
      class="emotion-0"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
